source/serial_api.c

```diff
@@ -470,4 +500,238 @@ void serial_break_clear(serial_t *obj)
+    handle->pTxBuffPtr = tx;
+    handle->TxXferSize = tx_length;
+    handle->TxXferCount = tx_length;
+
+    if(handle->State == HAL_UART_STATE_BUSY_RX) {
+        handle->State = HAL_UART_STATE_BUSY_TX_RX;
+    } else {
+        handle->State = HAL_UART_STATE_BUSY_TX;
+    }
+
+    // if the TX register is empty, directly input the first transmit byte
+    if (__HAL_UART_GET_FLAG(handle, UART_FLAG_TXE)) {
+        handle->Instance->DR = *handle->pTxBuffPtr++;
+        handle->TxXferCount--;
+    }
+    // chose either the tx reg empty or if last byte wait directly for tx complete
+    if (handle->TxXferCount != 0) {
+        handle->Instance->CR1 |= USART_CR1_TXEIE;
+    } else {
+        handle->Instance->CR1 |= USART_CR1_TCIE;
+    }
+
+    DEBUG_PRINTF("UART%u: Tx: 0=(%u, %u) %x\n", obj->serial.module+1, tx_length, tx_width, HAL_UART_GetState(handle));
+
+    return tx_length;
+}
+
+void serial_rx_asynch(serial_t *obj, void *rx, size_t rx_length, uint8_t rx_width, uint32_t handler, uint32_t event, uint8_t char_match, DMAUsage hint)
+{
+    // TODO: DMA usage is currently ignored
+    (void) hint;
+
+    bool use_rx = (rx != NULL && rx_length > 0);
+    IRQn_Type irq_n = UartIRQs[obj->serial.module];
+
+    if (!use_rx || !irq_n)
+        return;
+
+    obj->rx_buff.buffer = rx;
+    obj->rx_buff.length = rx_length;
+    obj->rx_buff.pos    = 0;
+    obj->rx_buff.width  = rx_width;
+
+    obj->serial.event      = (obj->serial.event & ~SERIAL_EVENT_RX_MASK) | (event & SERIAL_EVENT_RX_MASK);
+    obj->serial.char_match = char_match;
+
+    UART_HandleTypeDef *handle = &UartHandle[obj->serial.module];
+    // register the thunking handler
+    vIRQ_SetVector(irq_n, handler);
+    vIRQ_EnableIRQ(irq_n);
+
+    // HAL_StatusTypeDef rc = HAL_UART_Receive_IT(handle, rx, rx_length);
+
+    handle->pRxBuffPtr = rx;
+    handle->RxXferSize = rx_length;
+    handle->RxXferCount = rx_length;
+
+    if(handle->State == HAL_UART_STATE_BUSY_TX) {
+        handle->State = HAL_UART_STATE_BUSY_TX_RX;
+    } else {
+        handle->State = HAL_UART_STATE_BUSY_RX;
+    }
+
+    __HAL_UART_CLEAR_PEFLAG(handle);
+    handle->Instance->CR1 |= USART_CR1_RXNEIE | USART_CR1_PEIE;
+    handle->Instance->CR3 |= USART_CR3_EIE;
+
+    DEBUG_PRINTF("UART%u: Rx: 0=(%u, %u, %u) %x\n", obj->serial.module+1, rx_length, rx_width, char_match, HAL_UART_GetState(handle));
+}
+
+int serial_irq_handler_asynch(serial_t *obj)
+{
+    UART_HandleTypeDef *handle = &UartHandle[obj->serial.module];
+
+    int status = handle->Instance->SR;
+    int data = handle->Instance->DR;
+    int event = 0;
+
+    if (status & USART_SR_PE) {
+        event |= SERIAL_EVENT_RX_PARITY_ERROR;
+    }
+    if (status & (USART_SR_NE | USART_SR_FE)) {
+        event |= SERIAL_EVENT_RX_FRAMING_ERROR;
+    }
+    if (status & USART_SR_ORE) {
+        event |= SERIAL_EVENT_RX_OVERRUN_ERROR;
+    }
+
+    if ((status & USART_SR_TC) && (handle->State & 0x10)) {
```

Comment:
Not sure if I'm reading this wrong but if the UART set status TC and TXE then it will be treated as everything has been transmitted even if handler->TxXferCount != 0.
Then handle->State change to READY/BUSY_RX and will nerver enter this code again when TC is active and handler->TxXferCount == 0.